04_documentation/code_documentation_tools_demo.md

```diff
@@ -0,0 +1,302 @@
+
+       The `index.rst` file contains the main page of our homepage.
+
+       The `Makefile` and `make.bat` are there to trigger building the website.
+
+    4. Inspect website locally. Build the website
+
+       ```bash
+       make html
+       ```
+
+       and afterwards open `<prefix>/docs/build/html/index.html`. We can now also edit `index.rst` and rebuild to see the changes.
+
+       Run make once without argument to show all the target options.
+
+       ```bash
+       make
+       ```
+
+    5. Add another subpage, e.g., "Tutorials".
+
+       Add directory `tutorials/` with files `overview.rst` and `tutorial_case1.rst` (maybe `tutorial_case2.rst`).
+
+       Add content to files:
+
+       `tutorial_case1.rst`: Add headline, some text and two lower level headers.
+
+       ```rst
+       Tutorial 1
+       ==========
+
+       Tutorial description goes here.
+
+       Step 1
+       ------
+
+       Description of step 1.
+
+       Step 2
+       ------
+
+       Description of step 2.
+       ```
+
+       `overview.rst`: Add headline and a table of contents (TOC). It will be parsed from the `index.rst` such that the main header of the tutorial files will be present on the main page.
+
+       ```rst
+       Tutorial overview
+       =================
+
+       .. toctree::
+          :maxdepth: 2
+          :caption: tutorials:
+
+          tutorial_case1.rst
+          tutorial_case2.rst
+       ```
+
+       **Note** One must be careful with the indentation!
+    6. We can include the `REAMDE.rst` into the `index.rst`
+
+       ```diff
+       +.. include:: ../README.rst
+       ```
+
+       **Note**: In the git repository this addition currently happens in the next step.
+
+## 3. Sphinx: Code Documentation
+
+- Include our source files to `conf.py`. To do so we uncomment the lines
+
+  ```diff
+  -#import os
+  -#import sys
+  -#sys.path.insert(0, os.path.abspath('.'))
+  +import os
+  +import sys
+  +sys.path.insert(0, os.path.abspath('.'))
+  ```
+
+  and enable the autodoc extenstion
+
+  ```diff
+  -extensions = [
+  +extensions = ['sphinx.ext.autodoc'
+  ```
+
+- Create autodoc template files
+
+  ```bash
+  sphinx-apidoc -o source ../src/
+
+  Creating file source/solver.rst.
+  Creating file source/modules.rst.
+  ```
+
+  We can inspect the generated files. The files do not contain the actual documentation, but rather a template. The actual documentation is generated once `sphinx` is invoked.
+- Recreate website with `make html`. If there is an error message about "consistency" build the website again and/or comment out `from output` from the `diffusionsolver.py` file. Afterwards the error disappeared for me.
+    - We now can check out "Index" or "Module Index" and will find some documentation.
+- Add additional documentation to `diffusionsolver.py`. Example for `class DiffusionSolver` (should be added to constructire `__init__`)
```

Comment:
constructire?

Reply:
"constructor" if one cannot type.

Reply:
This comes later/is only common for multi-line docstrings and special doctring formatting standards like the ones from Google or Sphinx. I wanted to mention this here, but not enforce it yet. It will be more obvious/important when we reformat the docstrings.